frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/pushdialog/GitPushDialog.java

```diff
@@ -0,0 +1,358 @@
+      // cancel push and leave the push dialog open
+    }
+  }
+
+  @UIEffect
+  public PrePushHandler.Result runPrePushHandlersInModalTask() {
+    FileDocumentManager.getInstance().saveAllDocuments();
+    AtomicReference<PrePushHandler.Result> result = new AtomicReference<>(PrePushHandler.Result.OK);
+    new Task.Modal(project, /* title */ "Checking Commits...", /* canBeCancelled */ true) {
+      @Override
+      public void run(ProgressIndicator indicator) {
+        result.set(pushController.executeHandlers(indicator));
+      }
+
+      @Override
+      @UIEffect
+      public void onThrowable(Throwable error) {
+        if (error instanceof PushController.HandlerException) {
+          PushController.HandlerException handlerException = (PushController.HandlerException) error;
+          Throwable cause = handlerException.getCause();
+
+          String failedHandler = handlerException.getFailedHandlerName();
+          java.util.List<String> skippedHandlers = handlerException.getSkippedHandlers();
+
+          String suggestionMessageProblem;
+          if (cause instanceof ProcessCanceledException) {
+            suggestionMessageProblem = "${failedHandler} has been cancelled.";
+          } else {
+            // PushController.HandlerException constructor guarantees that cause is not null
+            assert cause != null : "Exception cause is null";
+            super.onThrowable(cause);
+            suggestionMessageProblem = "${failedHandler} has failed. See log for more details.";
+          }
+
+          String suggestionMessageQuestion = skippedHandlers.isEmpty()
+              ? "Would you like to push anyway or cancel the push completely?"
+              : "Would you like to skip all remaining pre-push steps and push, or cancel the push completely?";
+
+          suggestToSkipOrPush(suggestionMessageProblem + System.lineSeparator() + suggestionMessageQuestion);
+        } else {
+          super.onThrowable(error);
+        }
+      }
+
+      @Override
+      @UIEffect
+      public void onCancel() {
+        super.onCancel();
+        suggestToSkipOrPush("Would you like to skip all pre-push steps and push, or cancel the push completely?");
+      }
+
+      @UIEffect
+      private void suggestToSkipOrPush(String message) {
+        if (Messages.showOkCancelDialog(project,
+            message,
+            /* title */ "Push",
+            /* okText */ "Push Anyway",
+            /* cancelText */ "Cancel",
+            UIUtil.getWarningIcon()) == Messages.OK) {
+          result.set(PrePushHandler.Result.OK);
+        } else {
+          result.set(PrePushHandler.Result.ABORT);
+        }
+      }
+    }.queue();
+
+    PrePushHandler.Result resultValue = result.get();
+    assert resultValue != null : "Result value is null";
+    return resultValue;
+  }
+
+  @UIEffect
+  public void updateOkActions(@UnknownInitialization GitPushDialog this) {
+    if (pushAction != null) {
```

Comment:
Also, after `@UnknownInitialization` is removed from `this`, similar `if`s should no longer be needed